--- rc.lua	2015-09-10 00:18:58.806979497 +0200
+++ rc-test.lua	2015-09-10 00:31:21.956530857 +0200
@@ -452,6 +452,8 @@
                      raise = true,
                      keys = clientkeys,
                      buttons = clientbuttons } },
+    { rule = { class = "mpv" },
+      properties = { floating = false } },
     { rule = { class = "MPlayer" },
       properties = { floating = true } },
     { rule = { class = "pinentry" },
@@ -545,6 +547,7 @@
 
 --{{{ Autostart
 run_once("xscreensaver -no-splash")
+run_once("cairo-compmgr")
 --}}}
 --
 --vim: ft=lua tw=4
