--- rc.lua	2015-09-10 02:28:39.473687737 +0200
+++ rc-test.lua	2015-09-10 02:30:09.381567588 +0200
@@ -439,7 +439,7 @@
     mytasklist[s] = awful.widget.tasklist(s, awful.widget.tasklist.filter.currenttags, mytasklist.buttons)
 
     -- Create the wibox
-    mywibox[s] = awful.wibox({ position = "top", screen = s })
+    mywibox[s] = awful.wibox({ position = "bottom", screen = s })
 
     -- Widgets that are aligned to the left
     local left_layout = wibox.layout.fixed.horizontal()
@@ -449,7 +449,10 @@
 
     -- Widgets that are aligned to the right
     local right_layout = wibox.layout.fixed.horizontal()
-    if s == 1 then right_layout:add(wibox.widget.systray()) end
+    if s == 1 then
+        right_layout:add(wibox.widget.systray())
+        right_layout:add(separator)
+    end
     right_layout:add(battery_widget)
     right_layout:add(separator)
     right_layout:add(alsawidget.text)
