--- rc.lua	2015-09-10 23:28:06.244898527 +0200
+++ rc-test.lua	2015-09-10 23:28:49.593910435 +0200
@@ -869,8 +869,8 @@
         vicious.force({ alsa_widget.bar })
         alsa_widget.notify()
     end),
-    awful.key({}, "Pause", function() raise_conky() end, function() lower_conky() end),
-    awful.key({}, "Insert", function() toggle_conky() end)
+    awful.key({}, "Insert", function() raise_conky() end, function() lower_conky() end),
+    awful.key({}, "Pause", function() toggle_conky() end)
 )
 
 clientkeys = awful.util.table.join(
