--- rc.lua	2015-09-10 02:31:40.469420195 +0200
+++ rc-test.lua	2015-09-10 02:39:37.524849358 +0200
@@ -308,6 +308,7 @@
 end, 5, alsawidget.channel) -- relatively high update time, use of keys/mouse will force update
 alsawidget.text = wibox.widget.textbox()
 alsawidget.text:set_text("Vol:")
+alsabox = wibox.layout.margin(alsawidget.bar, 2, 2, 5, 5)
 
 battery_widget = wibox.widget.textbox()
 function batteryInfo(adapter)
@@ -456,7 +457,7 @@
     right_layout:add(battery_widget)
     right_layout:add(separator)
     right_layout:add(alsawidget.text)
-    right_layout:add(alsawidget.bar)
+    right_layout:add(alsabox)
     right_layout:add(separator)
     right_layout:add(datewidget)
     right_layout:add(separator)
